test/test_handlers.py

```diff
@@ -240,3 +240,63 @@ def guide():
 
     x = handlers.seed(guide, 0)()
     assert x.shape == shape
+
+
+@pytest.mark.parametrize('intervene,observe,flip', [
+    (True, False, False),
+    (False, True, False),
+    (True, True, False),
+    (True, True, True),
+])
+def test_counterfactual_query(intervene, observe, flip):
+    # x -> y -> z -> w
+
+    sites = ["x", "y", "z", "w"]
+    observations = {"x": 1., "y": None, "z": 1., "w": 1.}
+    interventions = {"x": None, "y": 0., "z": 2., "w": 1.}
+
+    def model():
+        with handlers.seed(rng_seed=0):
+            x = numpyro.sample("x", dist.Normal(0, 1))
+            y = numpyro.sample("y", dist.Normal(x, 1))
+            z = numpyro.sample("z", dist.Normal(y, 1))
+            w = numpyro.sample("w", dist.Normal(z, 1))
+            return dict(x=x, y=y, z=z, w=w)
+
+    if not flip:
+        if intervene:
+            model = handlers.do(model, data=interventions)
+        if observe:
+            model = handlers.condition(model, param_map=observations)
+    elif flip and intervene and observe:
+        model = handlers.do(
+            handlers.condition(model, param_map=observations),
+            data=interventions)
+
+    tr = handlers.trace(model).get_trace()
+    actual_values = tr["_RETURN"]["value"]
```

Comment:
In NumPyro, there is not `_RETURN` site. You can use
```
with handlers.trace() as tr:
    actual_values = model()
```
to get both trace `tr` and the returned value `actual_values`.